### DISCUSSION TITLE
Turbo Modules Starter/Template Project

### DISCUSSION THREAD

**Discussion by alon7** [reactwg/react-native-new-architecture#33] [Libraries]
# Problem
No starter/template project shows how to create a TurboModule library.
It's unfortunate because codegen works well with typescript, and without a usable template, library creators won't have the tools to upgrade their implementation to the new architecture.

# Currently
react-native-builder-bob today creates libraries with react-native@0.63.4. It doesn't support turbo modules and the new architecture.

There are a lot of differences and upgrades that need to be done to support 0.68+ turbo modules. 

In addition, there is no good Turbo Module template/example project that showcases how to use them seamlessly, generate code, and create a library with typescript and shared c++ code. 

# Solution
For that reason, we at Nagish created [react-native-turbo-starter](https://github.com/talknagish/react-native-turbo-starter), which has the following features:
- [x] React Native 0.68+
- [x] Upgraded react-native-create-library module to TurboModule
- [x] Codegen with typescript
- [x] Objective-C using JSI without the bridge
- [x] Java/Kotlin using JSI without the bridge
- [x] Shared C++ TurboModule
- [x] Array/Dictionaries/Promises usage
- [x] Native usage (battery level)
- [x] M1 e2e Compilation of RN New Architecture
- [x] Hermes

Roadmap:
- [ ] Generate a new project (like bob, using npx) 
- [ ] Backward compatibility to support the bridge

# Feedback
The project has already gathered 280+ stars and interest from many people from the community. 

We would love to get feedback and see if there's an option to make it an official library generator for turbo modules

### reference 
[react-native-turbo-starter](https://github.com/talknagish/react-native-turbo-starter)
_Originally forked from [react-native-builder-bob](https://github.com/callstack/react-native-builder-bob)_

**Comment by thymikee** [comment]
Have you considered contributing this back to Bob? :)

**Reply by alon7** [reply]
We considered it, but bob is a large project without much documentation, and we thought it'd be overkill to try and implement it there. 
Is the benefit the ability to create a new project easily? 

Do you have any plans to update bob to turbo modules?

**Reply by thymikee** [reply]
> Originally forked from [react-native-builder-bob](https://github.com/callstack/react-native-builder-bob)

Um, since it's based on bob, I thought it's not that big of a deal to send a PR with improvements back? We'd like to add support for turbo modules and fabric there, but still don't have the timeline.

**Reply by alon7** [reply]
Yeah, it's derived from bob (we created the library using bob) but I think it'll be a bit too much work for us in the near future. 
But, I'm not as familiar with bob, what are the advantages of using it? (npx, popularity, and template engine, anything else?)

**Reply by thymikee** [reply]
Mainly popularity I think. Bob is a go-to tool when writing a native module/component currently and it would be great if the community of library creators wouldn't need to choose between various tools but have one official to go with. I thought it wouldn't be big of a deal if the lib was a fork of Bob, but if it's based on different ideas, then I'm not gonna push.
cc @satya164

**Reply by alon7** [reply]
I'll look into bob and try to evaluate the amount of work it is. Will update when we have a better idea

**Reply by thymikee** [reply]
Thank you!
